### PR TITLE
[core] Expose python type ray.ClusterID, and use it in to track actor and task last export time.

### PR DESCRIPTION
**Pull request by rynewang**
We had a field `_last_export_session_and_job` in `class RemoteFunction` and `class _ActorClassMetadata`. It's used to track whether this actor/task has been exported to GCS or not, without having to pay an RTT to GCS for `internal_kv_exists`. The problem is this approach is not robust: the "session" is not a real concept, it's just an incrementing int counter in the class; while the Job IDs are not unique among clusters - all clusters' first job are 01000000. This requires the field to be omitted in pickling which makes unnecessary exporting to GCS.

This PR uses `(cluster_id, job_id)` to track last export time, because the cluster ID is unique. To achieve this, we need to also expose the ClusterID from a pure C++ concept to also be a Python type.

### CODE REVIEW

**Review by jjyao — COMMENTED** [review]
lg

File: python/ray/includes/unique_ids.pxi [hunk 1]
```diff
@@ -343,6 +344,21 @@ cdef class ActorClassID(UniqueID):
     cdef CActorClassID native(self):
         return <CActorClassID>self.data
 
+cdef class ClusterID(UniqueID):
+
+    def __init__(self, id):
+        check_id(id)
+        self.data = CNodeID.FromBinary(<c_string>id)
```
Comment (jjyao) [inline python/ray/includes/unique_ids.pxi:351]: CNodeID?

File: python/ray/includes/unique_ids.pxi [hunk 2]
```diff
@@ -343,6 +344,21 @@ cdef class ActorClassID(UniqueID):
...
+
+    @classmethod
+    def from_hex(cls, hex_id):
+        binary_id = CNodeID.FromHex(<c_string>hex_id).Binary()
```
Comment (jjyao) [inline python/ray/includes/unique_ids.pxi:355]: ?

File: python/ray/_raylet.pyx [hunk 3]
```diff
@@ -2696,38 +2696,35 @@ cdef class GcsClient:
...
+            self.init_cluster_id = CClusterID.FromHex(cluster_id)
         self._connect(RayConfig.instance().py_gcs_connect_timeout_s())
 
     def _connect(self, timeout_s=None):
         cdef:
             int64_t timeout_ms = round(1000 * timeout_s) if timeout_s else -1
             size_t num_retries = self._nums_reconnect_retry
         with nogil:
-            status = self.inner.get().Connect(self.cluster_id, timeout_ms, num_retries)
+            status = self.inner.get().Connect(self.init_cluster_id, timeout_ms, num_retries)
```
Comment (jjyao) [inline python/ray/_raylet.pyx:2720]: I think the behavior is changed with this since we don't update `init_cluster_id`.

**Review by jjyao — APPROVED** [review]

**Comment by jjyao** [comment]
There is test failure.

File: python/ray/_private/node.py [hunk 4]
```diff
@@ -709,9 +709,8 @@ def _init_gcs_client(self):
                 gcs_address = self.gcs_address
                 client = GcsClient(
                     address=gcs_address,
-                    cluster_id=self._ray_params.cluster_id,
+                    cluster_id=self._ray_params.cluster_id,  # Hex string
                 )
-                self.cluster_id = client.get_cluster_id()
```
Comment (jjyao) [inline python/ray/_private/node.py:714]: There is a test failure related to this

Comment (rynewang) [reply python/ray/_private/node.py:714]: fixed

**Pull request merged**